File: src/2016/day8.zig
const std = @import("std");
const utils = @import("utils");
const Point2d = utils.Point2d(i32);

pub fn main() !void {
    // https://adventofcode.com/2016/day/8
    utils.printTitle(2016, 8, "Two-Factor Authentication");

    const m = utils.Monitor.init();
    defer m.deinit();

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    const input = comptime @embedFile("input");

    var s = Screen.init(50, 6, allocator);
    defer s.deinit();
    s.run(input);

    utils.printAnswer(1, s.countLit());
    utils.printAnswer(2, "");
    try s.print("  ");
}

const Screen = struct {
    data: std.AutoHashMap(Point2d, void),
    width: i32,
    height: i32,

    pub fn init(w: i32, h: i32, allocator: std.mem.Allocator) Screen {
        const map = std.AutoHashMap(Point2d, void).init(allocator);
        return Screen{ .data = map, .width = w, .height = h };
    }

    pub fn deinit(self: *@This()) void {
        self.data.deinit();
    }

    pub fn run(self: *@This(), input: []const u8) void {
        var it = std.mem.tokenizeAny(u8, input, "\r\n");
        while (it.next()) |line| {
            var p = utils.Parser.init(line);
            const firstToken = p.nextScalar(' ').?;

            if (std.mem.eql(u8, firstToken, "rect")) {
                self.rect(p.rest());
            } else if (std.mem.eql(u8, firstToken, "rotate")) {
                const rowOrColumn = p.nextScalar(' ').?;
                if (std.mem.eql(u8, rowOrColumn, "row")) {
                    self.rotateRow(p.rest());
                } else if (std.mem.eql(u8, rowOrColumn, "column")) {
                    self.rotateColumn(p.rest());
                }
            } else unreachable;
        }
    }

    fn rect(self: *@This(), input: []const u8) void {
        var p = utils.Parser.init(input);
        const w = std.fmt.parseInt(usize, p.nextScalar('x').?, 10) catch unreachable;
        const h = std.fmt.parseInt(usize, p.rest(), 10) catch unreachable;

        for (0..h) |y| {
            for (0..w) |x| {
                const pt = Point2d{ .x = @intCast(x), .y = @intCast(y) };
                self.data.put(pt, {}) catch unreachable;
            }
        }
    }

    fn rotateRight(comptime T: type, items: []T, ammount: usize) []T {
        const len = items.len;
        const effectiveCount = @mod(ammount, len);
        return items[(len - effectiveCount)..] ++ items[0..(len - effectiveCount)];
    }

    fn rotateRow(self: *@This(), input: []const u8) void {
        var p = utils.Parser.init(input);
        _ = p.nextScalar('=').?;
        const row = std.fmt.parseInt(i32, p.nextScalar(' ').?, 10) catch unreachable;
        _ = p.nextScalar(' ').?;
        const count = std.fmt.parseInt(i32, p.rest(), 10) catch unreachable;

        var buffer: [50]bool = undefined;
        std.debug.assert(buffer.len >= self.width);

        var x: i32 = 0;
        while (x != self.width) : (x += 1) {
            const pt = Point2d{ .x = x, .y = row };
            buffer[@intCast(x)] = self.data.contains(pt);
        }
        x = 0;
        while (x != self.width) : (x += 1) {
            const pt = Point2d{ .x = x + count, .y = row };
            self.setWrapping(pt, buffer[@intCast(x)]);
        }
    }

    fn rotateColumn(self: *@This(), input: []const u8) void {
        var p = utils.Parser.init(input);
        _ = p.nextScalar('=').?;
        const column = std.fmt.parseInt(i32, p.nextScalar(' ').?, 10) catch unreachable;
        _ = p.nextScalar(' ').?;
        const count = std.fmt.parseInt(i32, p.rest(), 10) catch unreachable;

        var buffer: [6]bool = undefined;
        std.debug.assert(buffer.len >= self.height);

        var y: i32 = 0;
        while (y != self.height) : (y += 1) {
            const pt = Point2d{ .x = column, .y = y };
            buffer[@intCast(y)] = self.data.contains(pt);
        }
        y = 0;
        while (y != self.height) : (y += 1) {
            const pt = Point2d{ .x = column, .y = y + count };
            self.setWrapping(pt, buffer[@intCast(y)]);
        }
    }

    fn setWrapping(self: *@This(), pt: Point2d, v: bool) void {
        const wrapped = pt.wrap(self.width, self.height);
        if (v) {
            self.data.put(wrapped, {}) catch unreachable;
        } else {
            _ = self.data.remove(wrapped);
        }
    }

    fn getChar(self: *const @This(), pt: Point2d) []const u8 {
        return if (self.data.contains(pt)) "\u{2588}" else " ";
    }

    pub fn print(self: *const @This(), indent: []const u8) !void {
        var y: i32 = 0;

        var stdout_buffer: [1024]u8 = undefined;
        var stdout_writer = std.fs.File.stdout().writer(&stdout_buffer);
        const stdout = &stdout_writer.interface;

        while (y != self.height) : (y += 1) {
            try stdout.print("{s}", .{indent});
            var x: i32 = 0;
            while (x != self.width) : (x += 1) {
                const pt = Point2d{ .x = x, .y = y };
                try stdout.print("{s}", .{self.getChar(pt)});
            }

            try stdout.print("\n", .{});
        }

        try stdout.flush();
    }

    pub fn countLit(self: *@This()) i32 {
        return @intCast(self.data.count());
    }
};

test "parts 1,2" {
    const example =
        \\rect 3x2
        \\rotate column x=1 by 1
        \\rotate row y=0 by 4
        \\rotate column x=1 by 1
    ;
    var s = Screen.init(7, 3, std.testing.allocator);
    defer s.deinit();
    s.run(example);
    try std.testing.expectEqual(6, s.countLit());
}

File: src/2016/day8.rs
fn main() {
    // https://adventofcode.com/2016/day/8
    println!("Day 8, 2016: Two-Factor Authentication");

    debug_assert_eq!(6, run(get_example(), 7, 3).count_lit());
    // debug_assert_eq!(0, part2(get_example()));

    let screen = run(get_input(), 50, 6);
    let p1 = screen.count_lit();
    println!("  Part 1: {}", p1);
    assert_eq!(106, p1);

    println!("  Part 2:");
    screen.print("  ");
    // assert_eq!(0, p2);
}

struct Screen {
    lines: Vec<Vec<char>>,
    height: u16,
}

const LIT: char = '\u{2588}';

impl Screen {
    fn new(width: u16, height: u16) -> Self {
        Screen {
            lines: vec![vec![' '; width.into()]; height.into()],
            height: height,
        }
    }

    fn run(&mut self, instructions: &str) {
        for line in instructions.lines() {
            if line.starts_with("rotate row y=") {
                self.rotate_row(&line[13..]);
            } else if line.starts_with("rotate column x=") {
                self.rotate_col(&line[16..]);
            } else if line.starts_with("rect ") {
                self.rect(&line[5..]);
            }
        }
    }

    fn rect(&mut self, line: &str) {
        let mut parts = line.split("x");
        let w: i32 = parts.next().unwrap().parse().unwrap();
        let h: i32 = parts.next().unwrap().parse().unwrap();

        for y in 0..h {
            for x in 0..w {
                self.lines[y as usize][x as usize] = LIT;
            }
        }
    }

    fn rotate_row(&mut self, line: &str) {
        let mut parts = line.split(" by ");
        let y: usize = parts.next().unwrap().parse().unwrap();
        let count: usize = parts.next().unwrap().parse().unwrap();
        self.lines[y].rotate_right(count);
    }

    fn rotate_col(&mut self, line: &str) {
        let mut parts = line.split(" by ");
        let x: usize = parts.next().unwrap().parse().unwrap();
        let count: usize = parts.next().unwrap().parse().unwrap();
        let mut column: Vec<_> = self.lines.iter().map(|l| l[x]).collect();
        column.rotate_right(count);

        for y in 0..self.height {
            self.lines[y as usize][x] = column[y as usize];
        }
    }

    fn print(&self, indent: &str) {
        for line in &self.lines {
            eprintln!("{}{}", indent, line.into_iter().collect::<String>());
        }
    }

    fn count_lit(&self) -> u64 {
        self.lines
            .iter()
            .flat_map(|l| l.iter())
            .filter(|&&c| c == LIT)
            .count() as u64
    }
}

fn run(input: &'static str, w: u16, h: u16) -> Screen {
    let mut screen = Screen::new(w, h);
    screen.run(input);
    screen
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2016/day8.txt").trim_end()
}

fn get_example() -> &'static str {
    "rect 3x2
rotate column x=1 by 1
rotate row y=0 by 4
rotate column x=1 by 1"
}
